### DISCUSSION TITLE
How do I support cancellation of kotlin coroutines on connection close for all REST requests

### DISCUSSION THREAD

**Discussion by calebkiage** [quarkusio/quarkus#41696] [Q&A]
I just started testing out Quarkus with Kotlin suspend handlers and it seems to work well. However, I noticed that these handlers don't support cancellation on request connection closing. I found a workaround, but it seems very cumbersome and would like to find out if there's a better way.

In the sample code below, endpoint `/long/a` cancels the delay by throwing the `CancellationException` when the connection is closed, but `/long/b` does not.

```kotlin
@Path("/long")
class SampleResource(private val log: Logger) {
    @GET
    @Path("a")
    @Produces(MediaType.TEXT_PLAIN)
    suspend fun longRunning(rc: RoutingContext): String = coroutineScope {
        val requestJob = Job(coroutineContext[Job])
        // Listen for the connection close event
        rc.request().connection().closeHandler {
            requestJob.cancel(CancellationException("Client connection closed"))
        }
        // Launch a coroutine within this job
        async(requestJob) {
            delay(10000)
            log.info("/long/a responding")
            "Completed request"
        }.await()
    }

    @GET
    @Path("b")
    @Produces(MediaType.TEXT_PLAIN)
    suspend fun hello(): String = coroutineScope {
        delay(10000)
        log.info("/long/b responding")
        "Completed request"
    }
}
```

```log
2024-07-04 17:30:07,579 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-1) HTTP Request to /long/a failed, error id: 1e9c93da-0f36-484d-8557-240fba83db7f-3: java.util.concurrent.CancellationException: Client connection closed
	at ...

2024-07-04 17:30:22,693 INFO  [org.acm.SampleResource] (vert.x-eventloop-thread-0) /long/b responding
```

[a comment by an automated account is not shown]

**Reply by geoand** [reply]
Doesn't it cancel automatically after 60 seconds or so?

**Reply by calebkiage** [reply]
I'm not sure. I'd have to test this and get back to you.
I thought to use this technique to make my API more efficient since in this case the server keeps using up resources for a request whose response won't be read.

**Reply by calebkiage** [reply]
I've just run a test with a 70s delay. I made a request on the client and immediately closed the connection. On the server side, after the 70s, the handler ran and completed.

**Reply by geoand** [reply]
Interesting, thanks for checking. Please open an issue with a sample application so we can investigate further.

**Reply by calebkiage** [reply]
Done. Is there a way to create a HTTP interceptor that would achieve the effect in the `longRunning` handler above?

**Reply by geoand** [reply]
Unlikely, but I need to check further.

**Reply by geoand** [reply]
Can you open a new issue for the client?

Thanks

**Reply by calebkiage** [reply]
Done.

**Comment by calebkiage** [comment]
This is confirmed to be working in 3.13.0.CR1. Thanks @geoand

**Reply by geoand** [reply]
🙏🏼
